tools/project_manager/diff.php

```diff
@@ -138,11 +141,25 @@
     echo "| <a href='$format_url'>$format_label</a>\n";
 }
 
+if (user_can_mentor_in_any_round()) {
+    // show option to change diff style
+    if ($bb_diffs) {
+        $bb_diffs_label = _("Show table-style diff");
+        $_GET["bb_diffs"] = "off";
+    } else {
+        $bb_diffs_label = _("Show BBCode-style diff");
+        $_GET["bb_diffs"] = "on";
+    }
+    $bb_diffs_url = "?" . attr_safe(http_build_query($_GET));
+    echo "| <a href='$bb_diffs_url'>$bb_diffs_label</a>\n";
+}
+
 echo "</p>\n";
 
 // ---------------------------------------------------------
 
-$diffEngine = new DifferenceEngineWrapper();
+$diffurl = "$code_url/tools/project_manager/diff.php?project=$projectid&amp;image=$image&amp;L_round_num=$L_round_num&amp;R_round_num=$R_round_num";
```

Comment:
`$image_url` and `$url` in this file use `&amp;` and I followed blindly. If you think the context argument means that this url should be treated differently, I can change it. I'm assuming you mean that I would pass the unencoded version into the constructor, then encode it later when I'm writing it to the webpage?

Reply:
My concern is this code doesn't know in what context this string will be used. If it is being output as HTML somewhere, that is where the encoding should be. For all we know at this point, this string may be shoved in a SQL statement, in which case its encoding would need to be different. So either we encode when using it or name it so the encoding is clear to consumers. I think $image_url and $url encode premeptively above because they are used at least within the same file, but even that I would think we should do differently to be honest.

Reply:
Thanks - I hadn't considered that before you raised it